src/hotspot/share/oops/symbolHandle.hpp

```diff
@@ -76,6 +110,23 @@ class SymbolHandleBase : public StackObj {
     Symbol::maybe_decrement_refcount(_temp);
   }
 
+  // Keep this symbol alive for some time to allow for reuse.
+  // Temp symbols for the same string can often be created in quick succession,
+  // and this queue allows them to be reused instead of churning.
+  void add_to_cleanup_delay_queue(Symbol* sym) {
+    TempSymbolDelayQueueNode* node = new TempSymbolDelayQueueNode(sym);
+    _cleanup_delay.push(*node);
+
+    // If the queue is now full, implement a one-in, one-out policy.
+    if (Atomic::add(&_cleanup_delay_len, 1, memory_order_relaxed) > _cleanup_delay_max_entries) {
```

Comment:
Why is incrementing relaxed?  Now I have to think hard about whether there
might be any ordering problems resulting from that.

Reply:
A colleague suggested that was the appropriate constraint ("there is no transitive memory effects that ride on the counter"), but I don't understand it enough to defend it myself. Atomic::inc uses memory_order_conservative. Happy to go with whatever is preferred.

Reply:
One of the things I'm worried about is that I think the decrement operation
might be able to become temporarily negative. I'm not sure whether that's a
problem. Maybe not. I was a little surprised that the queue length stuff is
signed rather than unsigned, as I expected the latter.

The idiom I've usually used for such things is increment-before-push and
decrement-after-pop. That ensures the counter never goes negative or
underflows.  Also usually use unsigned types for "sizes".